### PR TITLE
[Enhancement] Support litehrnet for ncnn

### PR DESCRIPTION
**Pull request by hanrui1sensetime**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily receiving feedbacks. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Support LiteHRNet for ncnn. The performance of AP is 0.663, AR is 0.728.

## Modification

- [x] Rewrite chunk for ncnn.
- [x] Register dummy adaptive_average_pooling for ncnn.
- [x] Add docstrings and unittests
- [x] performance test. 

## BC-breaking (Optional)

None.

## Use cases (Optional)

None.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a dependency on downstream projects of a newer version, this PR should be tested with all supported versions of downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmdeploy/codebase/mmpose/core/ops/adaptive_avg_pool_ncnn.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import torch
+
+
+class NcnnAdaptiveAvgPoolingOp(torch.autograd.Function):
```
Comment (grimoire) [inline mmdeploy/codebase/mmpose/core/ops/adaptive_avg_pool_ncnn.py:5]: Why don't we add a symbolic rewriter in `mmdeploy/python`?

File: mmdeploy/codebase/mmpose/models/backbones/litehrnet.py [hunk 2]
```diff
@@ -27,3 +27,122 @@ def cross_resolution_weighting__forward(ctx, self, x):
...
+    ]
+    return out
+
+
+@FUNCTION_REWRITER.register_rewriter(
+    'mmpose.models.backbones.litehrnet.ShuffleUnit.forward', backend='ncnn')
+def shuffle_unit__forward__ncnn(ctx, self, x):
+    """Rewrite ``forward`` for ncnn backend.
+
+    Rewrite this function to support export ``chunk``.
```
Comment (grimoire) [inline mmdeploy/codebase/mmpose/models/backbones/litehrnet.py:68]: Can we rewrite the chunk with split? It is more general.

File: mmdeploy/codebase/mmpose/models/backbones/litehrnet.py [hunk 3]
```diff
@@ -27,3 +27,122 @@ def cross_resolution_weighting__forward(ctx, self, x):
...
+    Rewrite this function to support export ``adaptive_avg_pool2d``
+    for ncnn. Ncnn has its own implement of adaptive average pooling.
+
+    Args:
+        x (list): block input.
+    """
+    from mmdeploy.codebase.mmpose.core.ops import \
+        ncnn_adaptive_avg_pool_forward
+
+    mini_size = torch.tensor([int(_) for _ in x[-1].shape[-2:]])
```
Comment (grimoire) [inline mmdeploy/codebase/mmpose/models/backbones/litehrnet.py:47]: `torch.tensor(x[-1].shape[-2:])`.

File: mmdeploy/codebase/mmpose/models/backbones/litehrnet.py (hunk 2 above)
Comment (hanrui1sensetime) [reply mmdeploy/codebase/mmpose/models/backbones/litehrnet.py:68]: > Can we rewrite the chunk with split? It is more general. If it did support, does it means that we do not need all rewriters below?

Because chunk has a parameter `axis`, which may lead `if` statement into the function, may looks ugly.

File: mmdeploy/codebase/mmpose/core/ops/adaptive_avg_pool_ncnn.py (hunk 1 above)
Comment (hanrui1sensetime) [reply mmdeploy/codebase/mmpose/core/ops/adaptive_avg_pool_ncnn.py:5]: > Why don't we add a symbolic rewriter in `mmdeploy/python`?

So what about `DetectionOutput` and `PriorBox` op in mmdet rewriter?

Comment (grimoire) [reply mmdeploy/codebase/mmpose/core/ops/adaptive_avg_pool_ncnn.py:5]: adaptive_avg_pool is a buildin function in PyTorch, if we add symbolic in `mmdeploy/python`, we can reuse it in other repo.
`DetectionOutput` and `PriorBox` seen not that general.

File: mmdeploy/codebase/mmpose/models/backbones/litehrnet.py (hunk 2 above)
Comment (grimoire) [reply mmdeploy/codebase/mmpose/models/backbones/litehrnet.py:68]: I don't get it, could you provide an example?

Comment (hanrui1sensetime) [reply mmdeploy/codebase/mmpose/models/backbones/litehrnet.py:68]: Ok, now I have tested using torch.index_select can export model correctly, so we can rewrite torch.chunk derectly.

File: mmdeploy/python/adaptive_avg_pool_ncnn.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import torch
+
+
+class NcnnAdaptiveAvgPoolingOp(torch.autograd.Function):
```
Comment (grimoire) [inline mmdeploy/python/adaptive_avg_pool_ncnn.py:5]: Errr, this is not what I means, there is an symbolic rewrite of adaptive_avg_pool here https://github.com/open-mmlab/mmdeploy/blob/master/mmdeploy/pytorch/ops/adaptive_avg_pool.py . You can add another rewrite for ncnn.

Comment (hanrui1sensetime) [reply mmdeploy/python/adaptive_avg_pool_ncnn.py:5]: > Errr, this is not what I means, there is an symbolic rewrite of adaptive_avg_pool here https://github.com/open-mmlab/mmdeploy/blob/master/mmdeploy/pytorch/ops/adaptive_avg_pool.py . You can add another rewrite for ncnn.

Ah, got it.

File: mmdeploy/pytorch/ops/adaptive_avg_pool.py [hunk 5]
```diff
@@ -78,3 +79,30 @@ def adaptive_avg_pool3d__default(ctx, *args):
     Align symbolic of adaptive_pool between different torch version.
     """
     return adaptive_avg_pool3d(*args)
+
+
```
Comment (grimoire) [inline mmdeploy/pytorch/ops/adaptive_avg_pool.py:83]: Why don't we use 
```python
@SYMBOLIC_REWRITER.register_symbolic('adaptive_avg_pool2d', backend='ncnn', is_pytorch=True)
```
just like the symbolic above?
We can use `torch.nn.functional.adaptive_avg_pool2d` instead of `NcnnAdaptiveAvgPoolingOp.apply`

**Comment by RunningLeon** [comment]
@hanrui1sensetime Could fix CI error

**Comment by tpoisonooo** [comment]
LGTM.

**Review by RunningLeon — APPROVED** [review]
LGTM

**Review by grimoire — APPROVED** [review]
LGTM

**Comment by XiaoYuhao** [comment]
I try to create the litehrnet ncnn model, but it still has some warnings:
`
WARNING: The shape inference of mmdeploy::AdaptiveAvgPool2d type is missing, so it may result in wrong shape inference for the exported graph. Please consider adding it in symbolic function.
WARNING: The shape inference of mmdeploy::AdaptiveAvgPool2d type is missing, so it may result in wrong shape inference for the exported graph. Please consider adding it in symbolic function.
......
`

I ingore this problem and try to run the litehrnet ncnn model. Then I get a runtime error. I want to kown how to fix this issue.

Appreciate a lot If you could provide some insights.
